perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -3239,18 +3240,33 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
 
 	@Override
 	public List<Attribute> getRequiredAttributes(PerunSession sess, Service service, Resource resource, Group group) throws InternalErrorException {
+		return getRequiredAttributes(sess, Collections.singletonList(service), resource, group);
+	}
+
+	@Override
+	public List<Attribute> getRequiredAttributes(PerunSession sess, List<Service> services, Resource resource, Group group) throws InternalErrorException {
 		if(!CacheManager.isCacheDisabled() && !perun.getCacheManager().wasCacheUpdatedInTransaction()) {
-			List<Integer> attrIds = getRequiredAttributeIds(service);
-			List<Attribute> attrs = perun.getCacheManager().getAttributesByIds(attrIds, new Holder(group.getId(), Holder.HolderType.GROUP), new Holder(resource.getId(), Holder.HolderType.RESOURCE));
+			List<Integer> attrIds = new ArrayList<>();
+			for (Service service : services) {
+				attrIds.addAll(getRequiredAttributeIds(service));
```

Comment:
We should make sure, that attribute IDs are unique, since they may overlap for multiple services.

Reply:
Done.